<a href="https://colab.research.google.com/github/blazaropinto/PDA_Data_Analysis_Python/blob/main/Sorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning 
---



In order to effectively analyse a dataset, often we need to prepare it first. 
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* Sort by a single column e.g.
  * `df.sort_values("Make") `
* Sort by multiple columns e.g. 
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make 
* Sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted* 

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score 
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [2]:
import pandas as pd

url = 'https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true'

happiness_original = pd.read_excel(url)
happiness_original.head()

Country          Region  ...  Generosity  Dystopia Residual
0  Switzerland  Western Europe  ...     0.29678            2.51738
1      Iceland  Western Europe  ...     0.43630            2.70201
2      Denmark  Western Europe  ...     0.34139            2.49204
3       Norway  Western Europe  ...     0.34699            2.46531
4       Canada   North America  ...     0.45811            2.45176

[5 rows x 12 columns]

##### sort the data by Happiness Score in ascending order:

In [3]:
happiness_HSordered = happiness_original.sort_values('Happiness Score')
happiness_HSordered.head()

Country                           Region  ...  Generosity  Dystopia Residual
157     Togo               Sub-Saharan Africa  ...     0.16681            1.56726
156  Burundi               Sub-Saharan Africa  ...     0.19727            1.83302
155    Syria  Middle East and Northern Africa  ...     0.47179            0.32858
154    Benin               Sub-Saharan Africa  ...     0.18260            1.63328
153   Rwanda               Sub-Saharan Africa  ...     0.22628            0.67042

[5 rows x 12 columns]

In [4]:
happiness_HSordered.tail()

Country          Region  ...  Generosity  Dystopia Residual
4       Canada   North America  ...     0.45811            2.45176
3       Norway  Western Europe  ...     0.34699            2.46531
2      Denmark  Western Europe  ...     0.34139            2.49204
1      Iceland  Western Europe  ...     0.43630            2.70201
0  Switzerland  Western Europe  ...     0.29678            2.51738

[5 rows x 12 columns]

In [5]:
print('The lowest score is',
      happiness_HSordered['Happiness Score'].iloc[0],
      ',',
      happiness_HSordered['Country'].iloc[0],
      '\nand the highest score,',
      happiness_HSordered['Happiness Score'].iloc[happiness_HSordered.shape[0]-1],
      ',',
      happiness_HSordered['Country'].iloc[happiness_HSordered.shape[0]-1])

The lowest score is 2.839 , Togo 
and the highest score, 7.587 , Switzerland


##### in the exercise above we see .iloc accounts for the position, but .loc accounts for the index "name" (even though it is an integer, it is the same as in the original df if the index is not dropped when ordering). That could be also done in this way:

In [6]:
happiness_HSordered2 = happiness_original.sort_values('Happiness Score', ignore_index=True)
happiness_HSordered2.head()   # look at index

Country                           Region  ...  Generosity  Dystopia Residual
0     Togo               Sub-Saharan Africa  ...     0.16681            1.56726
1  Burundi               Sub-Saharan Africa  ...     0.19727            1.83302
2    Syria  Middle East and Northern Africa  ...     0.47179            0.32858
3    Benin               Sub-Saharan Africa  ...     0.18260            1.63328
4   Rwanda               Sub-Saharan Africa  ...     0.22628            0.67042

[5 rows x 12 columns]

In [7]:
print('The lowest score is',
      happiness_HSordered2.loc[0,'Happiness Score'],
      ',',
      happiness_HSordered2.loc[0,'Country'],
      '\nand the highest score,',
      happiness_HSordered2.loc[happiness_HSordered.shape[0]-1, 'Happiness Score'],
      ',',
      happiness_HSordered2.loc[happiness_HSordered.shape[0]-1, 'Country'])

The lowest score is 2.839 , Togo 
and the highest score, 7.587 , Switzerland


### Exercise 2 - sort by multiple columns, display the first 5 rows 
---

1. Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order 
2. Display the first 5 rows of sorted data 

**Test output**:  
Records 122, 127, 147, 100, 96

In [8]:
happiness_HSordered.sort_values(['Economy (GDP per Capita)', 
                                'Health (Life Expectancy)'])[['Economy (GDP per Capita)', 
                                                              'Health (Life Expectancy)']].head()

Economy (GDP per Capita)  Health (Life Expectancy)
119                   0.00000                   0.09806
156                   0.01530                   0.22396
130                   0.01604                   0.22562
143                   0.06940                   0.29707
115                   0.07120                   0.34201

### Exercise 3 - sorting in descending order 
---
 
Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
43, 7, 144, 0, 3 Country and Region columns


In [9]:
happiness_HSordered.sort_values(['Freedom', 'Trust (Government Corruption)'], ascending=False)[[ 'Country', 'Region']].tail()

Country                           Region
136                  Angola               Sub-Saharan Africa
117                   Sudan               Sub-Saharan Africa
95   Bosnia and Herzegovina       Central and Eastern Europe
101                  Greece                   Western Europe
111                    Iraq  Middle East and Northern Africa

# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values 

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values 
---

1. Read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [13]:
url_london = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

df_london = pd.read_csv(url_london)
df_london.head()

code                  area  ... no_of_houses  borough_flag
0  E09000001        city of london  ...          NaN             1
1  E09000002  barking and dagenham  ...          NaN             1
2  E09000003                barnet  ...          NaN             1
3  E09000004                bexley  ...          NaN             1
4  E09000005                 brent  ...          NaN             1

[5 rows x 12 columns]

In [25]:
df_london.isnull().values.any()

True

In [26]:
df_london.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


##### `.values` converts the df into a np array

In [24]:
df_london.isnull().values

array([[False, False, False, ...,  True,  True, False],
       [False, False, False, ...,  True,  True, False],
       [False, False, False, ...,  True,  True, False],
       ...,
       [False, False, False, ...,  True,  True, False],
       [False, False, False, ...,  True,  True, False],
       [False, False, False, ...,  True,  True, False]])

##### `.any()` tells if there is any `True` in each column of the df

In [14]:
df_london.isnull().any()

code                 False
area                 False
date                 False
median_salary         True
life_satisfaction     True
mean_salary          False
recycling_pct         True
population_size       True
number_of_jobs        True
area_size             True
no_of_houses          True
borough_flag         False
dtype: bool

In [27]:
df_london.isnull().any().any() # same output as .any() used once on an array

True

both fuctions 'isnull()` and `isna()' are the same (https://pandas-docs.github.io/pandas-docs-travis/reference/api/pandas.DataFrame.isnull.html)

In [15]:
df_london.isna().any()  

code                 False
area                 False
date                 False
median_salary         True
life_satisfaction     True
mean_salary          False
recycling_pct         True
population_size       True
number_of_jobs        True
area_size             True
no_of_houses          True
borough_flag         False
dtype: bool

In [28]:
df_london.isna().sum()  

code                   0
area                   0
date                   0
median_salary         22
life_satisfaction    719
mean_salary            0
recycling_pct        211
population_size       53
number_of_jobs       140
area_size            405
no_of_houses         405
borough_flag           0
dtype: int64

### Exercise 5 - remove null values 
---

1. Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. Remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

##### Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)

In [31]:
part1 = df_london.dropna(subset=['life_satisfaction'])
print('Remaining rows:', part1.shape[0])

Remaining rows: 352


##### Remove all NA values across whole dataframe

In [32]:
part2 = df_london.dropna()
print('Remaining rows:', part2.shape[0])

Remaining rows: 267


## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [33]:
unique_entries = df_london.drop_duplicates()
display(unique_entries.shape,unique_entries.date.unique()) #???

(1071, 12)

array(['1999-12-01', '2000-12-01', '2001-12-01', '2002-12-01',
       '2003-12-01', '2004-12-01', '2005-12-01', '2006-12-01',
       '2007-12-01', '2008-12-01', '2009-12-01', '2010-12-01',
       '2011-12-01', '2012-12-01', '2013-12-01', '2014-12-01',
       '2015-12-01', '2016-12-01', '2017-12-01', '2018-12-01',
       '2019-12-01'], dtype=object)

In [34]:
unique_entries.head()

code                  area  ... no_of_houses  borough_flag
0  E09000001        city of london  ...          NaN             1
1  E09000002  barking and dagenham  ...          NaN             1
2  E09000003                barnet  ...          NaN             1
3  E09000004                bexley  ...          NaN             1
4  E09000005                 brent  ...          NaN             1

[5 rows x 12 columns]

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

In [61]:
ms_median = df_london['median_salary'].median()
upper_quartile = df_london['median_salary'].quantile(0.75)

def normalise_income(df):
  if df['median_salary'] < ms_median:
    return 1
  elif df['median_salary'] < upper_quartile:
    return 2
  else:
    return 3

##### prevent changing the original df by making a copy with `.copy()`

* *with `.copy()` an independent object instance will be created and the changes made to the new instance won't affect the first*
* *if we just assign the df to another name, both variables will reference the same object instance so any change to the new copy will affect the original*

In [63]:
normalized_london_df = df_london.copy()

In [65]:
normalized_london_df['median_salary'] = normalized_london_df.apply(normalise_income, axis=1)
normalized_london_df.head()

code                  area  ... no_of_houses  borough_flag
0  E09000001        city of london  ...          NaN             1
1  E09000002  barking and dagenham  ...          NaN             1
2  E09000003                barnet  ...          NaN             1
3  E09000004                bexley  ...          NaN             1
4  E09000005                 brent  ...          NaN             1

[5 rows x 12 columns]

##### To replace just values fulfilling one condition we can use `.mask()` 

In [67]:
london_df2 = df_london.copy()
london_df2 = london_df2.mask(london_df2['median_salary'] < ms_median, 1)
london_df2.head()


code            area        date  ...  area_size  no_of_houses borough_flag
0  E09000001  city of london  1999-12-01  ...        NaN           NaN            1
1          1               1           1  ...        1.0           1.0            1
2          1               1           1  ...        1.0           1.0            1
3          1               1           1  ...        1.0           1.0            1
4          1               1           1  ...        1.0           1.0            1

[5 rows x 12 columns]

##### To replace precise values we can use `.replace()`

In [69]:
london_df3 = df_london.copy()
london_df3.replace({'median_salary':{33020.0:1, 21480.0:2, 19568.0:3}}).head()

code                  area  ... no_of_houses  borough_flag
0  E09000001        city of london  ...          NaN             1
1  E09000002  barking and dagenham  ...          NaN             1
2  E09000003                barnet  ...          NaN             1
3  E09000004                bexley  ...          NaN             1
4  E09000005                 brent  ...          NaN             1

[5 rows x 12 columns]

### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [72]:
df_london.describe()

median_salary  life_satisfaction  ...  no_of_houses  borough_flag
count    1049.000000         352.000000  ...  6.660000e+02   1071.000000
mean    27977.792183           7.485057  ...  8.814682e+05      0.647059
std      6412.807487           0.198451  ...  3.690376e+06      0.478108
min     15684.000000           7.000000  ...  5.009000e+03      0.000000
25%     23857.000000           7.350000  ...  8.763550e+04      0.000000
50%     27441.000000           7.510000  ...  1.024020e+05      1.000000
75%     30932.000000           7.640000  ...  1.262760e+05      1.000000
max     61636.000000           7.960000  ...  2.417217e+07      1.000000

[8 rows x 7 columns]

In [83]:
df_london.number_of_jobs.median()

157000.0

In [84]:
def normalise_jobs(df):
  if df['number_of_jobs'] < 157000:
    return 1
  elif df['number_of_jobs'] < 35750000:
    return 2
  else:
    return 3

In [85]:
normalised_copy = df_london.copy()
normalised_copy['number_of_jobs'] = normalised_copy.apply(normalise_jobs, axis=1)
normalised_copy.head()

code                  area  ... no_of_houses  borough_flag
0  E09000001        city of london  ...          NaN             1
1  E09000002  barking and dagenham  ...          NaN             1
2  E09000003                barnet  ...          NaN             1
3  E09000004                bexley  ...          NaN             1
4  E09000005                 brent  ...          NaN             1

[5 rows x 12 columns]

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1 

In [88]:
lower_q = df_london.no_of_houses.quantile(0.25)
upper_q = df_london.no_of_houses.quantile(0.75)

def normalise_house(df):
  if df['no_of_houses'] > upper_q:
    return 3
  elif df['no_of_houses'] > lower_q:
    return 2
  else:
    return 1

df_house_normalized = df_london.copy()
df_house_normalized['housing_volume'] = df_house_normalized.apply(normalise_house, axis=1)
df_house_normalized.head()


code                  area  ... borough_flag  housing_volume
0  E09000001        city of london  ...            1               1
1  E09000002  barking and dagenham  ...            1               1
2  E09000003                barnet  ...            1               1
3  E09000004                bexley  ...            1               1
4  E09000005                 brent  ...            1               1

[5 rows x 13 columns]

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.


In [89]:
area_mean = df_london.area_size.mean()

def normalise_area(input):
  if input['area_size'] > area_mean:
    return 0
  else:
    return 1
  
df_london['area_size_normalised'] = df_london.apply(normalise_area, axis=1)
df_london.head()

code                  area  ... borough_flag  area_size_normalised
0  E09000001        city of london  ...            1                     1
1  E09000002  barking and dagenham  ...            1                     1
2  E09000003                barnet  ...            1                     1
3  E09000004                bexley  ...            1                     1
4  E09000005                 brent  ...            1                     1

[5 rows x 13 columns]

In [92]:
df_london[df_london.area_size_normalised == 0].head(1)

code     area  ... borough_flag  area_size_normalised
146  E92000001  england  ...            0                     0

[1 rows x 13 columns]